zanata-war/src/main/java/org/zanata/service/impl/MergeTranslationsWork.java

```diff
@@ -0,0 +1,137 @@
+    private final int batchStart;
+    private final int batchLength;
+    private final boolean useNewerTranslation;
+    private final List<HLocale> supportedLocales;
+
+    private final TextFlowDAO textFlowDAO;
+
+    private final TranslationStateCache translationStateCacheImpl;
+
+    private final Stopwatch stopwatch = Stopwatch.createUnstarted();
+
+    @Override
+    protected Integer work() throws Exception {
+        stopwatch.start();
+
+        List<HTextFlow[]> matches =
+                textFlowDAO.getSourceByMatchedContext(
+                        sourceVersionId, targetVersionId, batchStart,
+                        batchLength);
+
+        for (HTextFlow[] results : matches) {
+            HTextFlow sourceTf = results[0];
+            HTextFlow targetTf = results[1];
+            boolean hasChanged = false;
+
+            for (HLocale hLocale : supportedLocales) {
+                HTextFlowTarget sourceTft =
+                        sourceTf.getTargets().get(hLocale.getId());
+                // only process translated state
+                if (sourceTft == null || !sourceTft.getState().isTranslated()) {
+                    continue;
+                }
+
+                HTextFlowTarget targetTft =
+                        targetTf.getTargets().get(hLocale.getId());
+                if (targetTft == null) {
+                    targetTft = new HTextFlowTarget(targetTf, hLocale);
+                    targetTft.setVersionNum(0);
+                    targetTf.getTargets().put(hLocale.getId(), targetTft);
+                }
+
+                if (MergeTranslationsServiceImpl.shouldMerge(sourceTft,
+                        targetTft, useNewerTranslation)) {
+                    hasChanged = true;
+                    ContentState oldState = targetTft.getState();
+
+                    targetTft.setContents(sourceTft.getContents());
+                    targetTft.setState(sourceTft.getState());
+                    targetTft.setLastChanged(sourceTft.getLastChanged());
+                    targetTft.setLastModifiedBy(sourceTft.getLastModifiedBy());
+                    targetTft.setTranslator(sourceTft.getTranslator());
+
+                    if (sourceTft.getComment() == null) {
+                        targetTft.setComment(null);
+                    } else {
+                        HSimpleComment hComment = targetTft.getComment();
+                        if (hComment == null) {
+                            hComment = new HSimpleComment();
+                            targetTft.setComment(hComment);
+                        }
+                        hComment
+                                .setComment(sourceTft.getComment()
+                                    .getComment());
+                    }
+                    targetTft.setRevisionComment(MessageGenerator
+                            .getMergeTranslationMessage(sourceTft));
+
+                    raiseSuccessEvent(targetTft, oldState);
```

Comment:
Line 84-107 could easily be extracted to a method, since this is way too big.

Could probably also extract the whole of the HTextFlow block and the whole of the HLocale block to separate methods to make this a lot easier to follow. I should be able to see the whole of each method on my screen at once is possible.